### ISSUE TITLE
[Question] Does orleans collect grains when the silo is low on memory?

### ISSUE THREAD

**Issue by jamcar23** [dotnet/orleans#6843]
I know orleans collects grains after they've been idle for a certain amount of time but, does orleans stop referencing / collect old grains (which haven't become idle) when a silo is low on memory?

I'm asking because I'm running into a situation where a silo is crashing because it's running out of memory. After debugging, I've found the cause to be a large number of stateful grains which never get collected by the GC because they're still being referenced by the orleans runtime (because they haven't become idle yet).

Is this the intended behavior?

**Comment by sergeybykov** [comment]
Orleans does not deactivate grain based on memory pressure, only on the idle time basis. There are arguments to do that and there are arguments why that would potentially be masking issues.

If you'd like in such a situation to deactivate old grains, there are two possibilities. 
1) There are a bunch of old idle grains that are using memory for no good reason.
2) There are old grains that aren't idle (were invoked recently).

The right solution for 1) is to configure the idle time for those grains to a shorter period of time, so that they wouldn't be hanging around for so long. I'd argue that's much better than to wait until the app starts experiencing memory pressure.

Case 2) won't be helped by deactivating such grains in a memory pressure situation. If those grains are not idle, they will get reactivated again, which is more expensive than letting them stay in memory.

Bottom line. The 2 hour default idle time is too generous for most grain types. In my opinion, it is better to configure it per grain type according to the app scenario. I believe that should prevent most memory pressure situations from happening in the first place.

**Comment by jamcar23** [comment]
Okay, thanks for your insight. I'll look at changing the idle times.
